# Solving PCA manually using numpy

In [30]:
import numpy as np
import math

### Dataset for SBP and DBP

In [31]:
data = np.array([[126, 128, 128, 130, 130, 132], [78, 80, 82, 82, 84, 86]])
cols = ["SBP", "DBP"]

In [32]:
# Mean and Sample variance(n-1)
SBP_mean = data[0].mean()
DBP_mean = data[1].mean()
SBP_var = data[0].var(ddof=1)
DBP_var = data[1].var(ddof=1)

In [33]:
SBP_mean, SBP_var, DBP_mean, DBP_var

(np.float64(129.0), np.float64(4.4), np.float64(82.0), np.float64(8.0))

## Step:1 Centering the data (x-mean)

In [34]:
# Centered data (x-mean)
SBP_center = data[0]-SBP_mean
DBP_center = data[1]-DBP_mean

In [35]:
data = np.vstack((data, SBP_center, DBP_center))

In [36]:
data.T

array([[126.,  78.,  -3.,  -4.],
       [128.,  80.,  -1.,  -2.],
       [128.,  82.,  -1.,   0.],
       [130.,  82.,   1.,   0.],
       [130.,  84.,   1.,   2.],
       [132.,  86.,   3.,   4.]])

In [8]:
data_len = data.shape[1]
data_len

6

## Step:2 Finding the Covariance Matrix

In [9]:
covariance = (1/(data_len-1)*((SBP_center*DBP_center).sum()))
covariance

np.float64(5.6000000000000005)

In [10]:
cov_mat = np.array([[SBP_var, covariance], [covariance, DBP_var]])
cov_mat

array([[4.4, 5.6],
       [5.6, 8. ]])

## Step 3: Calculating Eigenvalues

$(A - \lambda I)\mathbf{v} = 0$

Finding quaratic equation of eigenvalues:

$\lambda ^ 2 - 12.4 \lambda + 3.84$

$x = \frac{-b \pm \sqrt{b^2 - 4ac}}{2a}$

$\lambda = \frac{-12.4 \pm \sqrt{12.4^2 - 4(1)(3.84)}}{2(1)}$

In [11]:
eigen_1 = (-(-12.4) + math.sqrt((12.4**2) - (4*1*3.84))) / (2*1)
eigen_2 = (-(-12.4) - math.sqrt((12.4**2) - (4*1*3.84))) / (2*1)

eigen_1, eigen_2

(12.082176467941098, 0.31782353205890246)

## Step 4: Finding the eigen vectors

$(A - \lambda I)\mathbf{v} = 0$

In [12]:
imat_1 = np.identity(2) * eigen_1
imat_2 = np.identity(2) * eigen_2
imat_1, imat_2

(array([[12.08217647,  0.        ],
        [ 0.        , 12.08217647]]),
 array([[0.31782353, 0.        ],
        [0.        , 0.31782353]]))

In [13]:
mat_vec1 = cov_mat - imat_1
mat_vec2 = cov_mat - imat_2
mat_vec1, mat_vec2

(array([[-7.68217647,  5.6       ],
        [ 5.6       , -4.08217647]]),
 array([[4.08217647, 5.6       ],
        [5.6       , 7.68217647]]))

### Finding the $(A- \lambda I) \mathbf{v}=0$

$\lambda _1$ = 12.08217647

$\lambda _2$ = 0.31782353

1. Solving $(A- \lambda _1 I)$:

$\begin{bmatrix}
  -7.68217647 &  5.6\\ 
  5.6 & -4.08217647
\end{bmatrix}$

2. Solving $(A- \lambda _2 I)$:

$\begin{bmatrix}
  4.08217647 &  5.6\\ 
  5.6 & 7.68217647
\end{bmatrix}$

1 and 2 are same but swapped diagonal elements

### Eigen Vector equation:

**Eigen Vector 1:**

$-7.68217647x + 5.6y = 0$

$5.6y = 7.68217647x$

$\begin{bmatrix} 7.682/5.6 \\ 1 \end{bmatrix}$

$\begin{bmatrix} 1.372 \\ 1 \end{bmatrix}$

**Eigen Vector 2:**

$4.08217647x + 5.6y = 0$

$5.6y = -4.08217647x$

$\begin{bmatrix} -4.082/5.6 \\ 1 \end{bmatrix}$

$\begin{bmatrix} -0.729 \\ 1 \end{bmatrix}$

In [14]:
mat_vec1, mat_vec2

(array([[-7.68217647,  5.6       ],
        [ 5.6       , -4.08217647]]),
 array([[4.08217647, 5.6       ],
        [5.6       , 7.68217647]]))

In [15]:
eigen_vec1 = np.array([mat_vec1[0][0] / mat_vec1[0][1],1])
eigen_vec2 = np.array([mat_vec2[0][0] / mat_vec2[0][1],1])
eigen_vec1, eigen_vec2

(array([-1.37181723,  1.        ]), array([0.72896008, 1.        ]))

## Step 5: Normalizing the eigen vectors

$\frac{1}{||v||} . v = \frac{x / ||v||}{y / ||v||}$

$||v|| = \sqrt{v_1 ^ 2 + v_2 ^ 2}$

In [16]:
vec1_dist = math.sqrt((eigen_vec1[0]**2) + (eigen_vec1[1]**2))
vec2_dist = math.sqrt((eigen_vec2[0]**2) + (eigen_vec2[1]**2))
vec1_dist, vec2_dist

(1.6976108219192407, 1.2374905266001552)

In [17]:
eigen_vec1 /= vec1_dist
eigen_vec2 /= vec2_dist
eigen_vec1, eigen_vec2

(array([-0.80808699,  0.58906316]), array([0.58906316, 0.80808699]))

In [18]:
eigen_vectors = np.vstack((eigen_vec1, eigen_vec2))
eigen_vectors

array([[-0.80808699,  0.58906316],
       [ 0.58906316,  0.80808699]])

In [19]:
# Built-in methods
eigenvalues, eigenvectors = np.linalg.eig(cov_mat)
eigenvalues, eigenvectors

(array([ 0.31782353, 12.08217647]),
 array([[-0.80808699, -0.58906316],
        [ 0.58906316, -0.80808699]]))

## Step 6: Principle Component Calculation

- Muplitplying Centered Data and Eigen Vectors

- Finding the variance of resultant matrix

- Which is Data Covered by Principle Component

In [20]:
center_data = data[2:].T
center_data

array([[-3., -4.],
       [-1., -2.],
       [-1.,  0.],
       [ 1.,  0.],
       [ 1.,  2.],
       [ 3.,  4.]])

In [ ]:
pc = np.dot(center_data, eigen_vectors)
pc1_var = pc[0].var(ddof=1)
pc2_var = pc[1].var(ddof=1)
pc

array([[ 0.06800833, -4.99953747],
       [-0.37003933, -2.20523715],
       [ 0.80808699, -0.58906316],
       [-0.80808699,  0.58906316],
       [ 0.37003933,  2.20523715],
       [-0.06800833,  4.99953747]])

In [22]:
pc1_var, pc2_var

(np.float64(12.840010200459021), np.float64(1.683975518898345))

## Step 7: Percentage Covered by PC

$pc_1 coverage  = \frac{pc_1}{pc_1 + pc_2}$

$pc_2 coverage  = \frac{pc_2}{pc_1 + pc_2}$

In [23]:
pc1_cover = pc1_var / (pc1_var + pc2_var)
pc2_cover = pc2_var / (pc1_var + pc2_var)

pc1_cover, pc2_cover

(np.float64(0.8840555511801443), np.float64(0.11594444881985567))

In [26]:
pca_col = max(pc1_cover, pc2_cover)

print(f"The PCA column is SBP with coverage of: {round(pca_col*100, 2)}%")

The PCA column is SBP with coverage of: 88.41%


In [ ]:
# Finding z-score manual method

num_data = data[num_cols].copy()
zscore_cols = ['Age_zscore', 'Blood_Pressure_zscore', 'Cholesterol_zscore']

num_data['Age_zscore'] = round((num_data['Age'] - num_data['Age'].mean()) / math.sqrt(num_data['Age'].var()), 2)
num_data['Blood_Pressure_zscore'] = round((num_data['Blood_Pressure'] - num_data['Blood_Pressure'].mean()) / math.sqrt(num_data['Blood_Pressure'].var()), 2)
num_data['Cholesterol_zscore'] = round((num_data['Cholesterol'] - num_data['Cholesterol'].mean()) / math.sqrt(num_data['Cholesterol'].var()), 2)

num_data

In [ ]:
# Finding IQR

Q1 = data['Age'].quantile(0.25)
Q3 = data['Age'].quantile(0.75)
IQR = Q3 - Q1
outliers = data[(data['Age'] < (Q1 - 1.5 * IQR)) | (data['Age'] > (Q3 + 1.5 * IQR))]
outliers